docs/spec/lightclient/detection.md

```diff
@@ -0,0 +1,491 @@
+   - *FullNodes \intersect Secondaries = {}*
+   - *FullNodes \intersect FaultyNodes = {}*
+   - *Secondaries \intersect FaultyNodes = {}*
+   
+and the following transition invariant
+   - *FullNodes' \union Secondaries' \union FaultyNodes' = FullNodes \union Secondaries \union FaultyNodes*
+
+## Solution
+
+
+
+### Inter Process Communication
+
+
+```go
+func FetchLightBlock(peer PeerID, height Height) LightBlock
+```
+See the [verification specification][verification] for details.
+
+### Auxiliary Functions (Local)
+
+
+```go
+Replace_Secondary(addr Address)
+```
+- Expected precondition
+    - *FullNodes* is nonempty
+- Expected postcondition
+    - addr is moved from *Secondaries* to *FaultyNodes*
+    - an address *a* is moved from *FullNodes* to *Secondaries*
+- Error condition
+    - if precondition is violated
+
+
+
+### From the verifier
+```go
+func VerifyToTarget(primary PeerID, lightStore LightStore,
+                    targetHeight Height) (LightStore, Result)
+```
+See the [verification specification][verification] for details.
+
+
+
+## Solution
+
+Shared data of the light client
+- a pool of full nodes *FullNodes* that have not been contacted before
+- peer set called *Secondaries*
+- primary
+- LightStore
+
+
+
+
+The problem is solved by calling  the function `ForkDetector` with
+a lightstore that contains a light block that has
+just been verified by the verifier. 
+
+
+```go
+func ForkDetector(ls LightStore)  {
+    Forks.Init // initialize a container in which we collect forks
+	for i, s range Secondaries {
+		sh := FetchLightBlock(s,LightStore.LatestVerified().Height)
+		// as the check below only needs the header, it is sufficient
+		// to download the header rather than the LighBlock
+		if LightStore.LatestVerified().Header == sh.Header {
+				// header matches. we do nothing.
+	    }
+		else {
+			    // [LCD-REQ-REP]
+			    // header does not match. there is a situation.
+				// we try to verify sh by querying s
+				// we set up an auxiliary lightstore with the highest
+			    // trusted lightblock and the lightblock we want to verify
+				auxLS.Init
+				auxLS.Update(LightStore.LatestTrusted(), StateVerified);
+				auxLS.Update(sh,StateUnverified);
+				result := VerifyToTarget(s, auxLS, LightStore.LatestVerified().Height)
+				if result = (_,ResultSuccess) || (_,EXPIRED) {
+				    // we verified header sh which is conflicting to hd
+					// there is a fork on the main blockchain.
+                    // If return code was EXPIRED it might be too late
+			    	// to punish, we still report it.
+					Forks.Add(sh)
+				}
+				else {
+					// s might be faulty or unreachable
+					Replace_peer(s)
+					// If a new node is added to secondaries, this
+					// should not imply an additional loop iteration.
+					// We assume one of the secondaries is correct.
+				}
+			}
+	}
+	if Forks.isEmpty {
+	    LightStore.Update(LightStore.LatestVerified(), StateTrusted)
+	}
+	return (LightStore,Forks,OK)
```

Comment:
I think that we are missing the main logic, i.e., describing what happens at the caller of ForkDetector. I guess in this case we report evidence and shut down light client?

Reply:
Yes, we need to clarify this. I will add a TODO in the beginning of the document.